### ISSUE TITLE
RAML nillable shorthand properties lose default value

### ISSUE THREAD

**Issue by kdubb** [aml-org/amf#869]
Your issue may already be reported! Please search on Github issues before creating one.

* **I'm submitting a ...**
  - [X] bug report
  - [ ] feature request
  - [ ] question
  - [ ] other

* **What is the current behavior?**

RAML nillable shorthand properties lose their default value.

* **If the current behavior is a bug, please provide the minimal steps to reproduce with AMF code. Please don't upload AMF code mixed with yours.**

For the following RAML, the `schema().defaultValue()` property of each `operation.request().headers()` has a default value except for the last parameter `def2`; which uses RAML nillable shorthand.
```raml
/tests:
  get:
    displayName: fetchTest
    headers:
      int:
        type: integer
        default: 1
      def1?:
        type: integer
        default: 5
      def2:
        type: integer?
        default: 10
    responses:
      200:
        body:
          type: Test

```

* **What is the expected behavior?**

The `defaultValue` should be available for all properties.

* **What is the motivation/use case for changing the behavior? (for feature requests)**

* **Please tell us about your environment:**
  
  - AMF Version: 4.7.0
  - AMF Distribution: JVM
  - Operating System: macOS
  - Link to your project: https://github.com/outfoxx/sunday-generator

* **Other information** (e.g. detailed explanation, stack traces, related issues, suggestions how to fix, links for us to have context, eg. StackOverflow, Gitter, etc)

**Comment by kdubb** [comment]
Note that in expanded union form... e.g.
```raml
def3:
  type: integer | nil
  default: 20
```

The default value is appropriately assigned.

**Comment by AgustinBettati** [comment]
Hello @kdubb! thanks for bringing this up. 
I created an issue in our internal tracking system for this: https://www.mulesoft.org/jira/browse/APIMF-2973.
I will keep you updated on any progress.

**Comment by kdubb** [comment]
@AgustinBettati I have found **numerous** issues with nillable shorthand properties. I haven't reported all of them because they basically seem to be just unsupported at the current time.

That being said, _please_ fix and support them! RAML is concise and this is one of the things that makes it concise and why we use it exclusively as our design language over OpenAPI.

**Comment by AgustinBettati** [comment]
Trailing question marks in properties and schemas are fully supported, what are exactly the issues you are encountering?

**Comment by kdubb** [comment]
As I said there have been numerous. I'll continue to report them as we re-encounter them.

Here is another problem we encounter. The following RAML produces the inexplicable error ` Property 'required' not supported in a RAML 1.0 string node`.

```raml
types:
  Test:
    properties:
      name:
        required: false
        type: string?
```

Edit: corrected invalid RAML for history's sake

**Comment by kdubb** [comment]
To be clear, any use of the shorthand precludes using an explicit `required: false|true`.

**Comment by kdubb** [comment]
Also, another similar issue. The following RAML, using both shorthands, produces the error `Property 'type?' not supported in a RAML 1.0 string node`.

```raml
types:
  Test:
    properties:
      name?: string?
```

Edit: corrected invalid RAML for history's sake

**Comment by kdubb** [comment]
@AgustinBettati Should I open separate issues for these issues?

**Comment by AgustinBettati** [comment]
I made some simple adjustments to both cases as the RAML definitions where not correctly defined.

**First case** - nillable type with required facet in property:
``` yaml
types:
  Test:
    properties:
      name:
        required: false
        type: string?
```
This is effectively giving an incorrect validation for the 'required' facet, please raise a separate issue! @kdubb
Note: the expanded form `string | nil` works correctly. 

**Second case** - optional property with nillable type:
``` yaml
types:
  Test:
    properties:
      name?: string?
```
This is parsing without any errors, the example you provided is simply incorrect.

**Comment by kdubb** [comment]
@AgustinBettati Yes, apologies for the hastily written and incorrect RAML.

I've opened #874 for the first case.

The second case (optional property with billable type) seems it might be an issue with our usage of the generated shapes; will debug and see if it warrants any issue.

**Comment by arielmirra** [comment]
Hi @kdubb! I've fixed this issue, it was a problem when dealing with the nullable notation '?'.
Now this raml:
```raml
def3:
  type: integer?
  default: 20
```

should be equivalent to:
```raml
def3:
  type: integer | nil
  default: 20
```

This fix will be present in the next AMF release approximately on August 23rd. It will also be available in AMF 5.0.0-beta.4 if you're considering switching to the new AMF major version.

Thanks!
